File: backend/python/model/scheme.py
from collections import defaultdict
from model.statement import Statement, LongStatement


class Scheme:
    def __init__(self, statements: list[LongStatement]):
        self.statements: dict[tuple[str, str], set[Statement]] = defaultdict(lambda: set())
        self.variables: set[str] = set() 
        self.order: dict[str, set[str]] = defaultdict(lambda: set())

        for statement in statements:
            self.variables.add(statement.variableFrom)
            self.variables.add(statement.variableTo)
            self.order[statement.variableFrom].add(statement.variableTo)
            self.statements[(statement.variableFrom, statement.variableTo)].add(Statement(statement.domain, statement.behaviour, statement.range))

        for k in self.variables:
            for i in self.variables:
                if k in self.order[i]:
                    self.order[i].update(self.order[k])
                    
    def __str__(self):
        return str(self.statements)

        

File: src/lib/types/scheme.ts
type VariableMap<T> = Map<string, Map<string, T>>;

export type Scheme = {
    statements: VariableMap<Statement[]>;
    variables:  Set<string>;
    order:      Map<string, Set<string>>; 
}

export type SchemeSerialised = {
    variables: string[];
    order: {
        variableFrom: string;
        variableTos:  string[];
    }[];
    statements: {
        variableFrom: string;
        variableTo:   string;
        statements:   Statement[];
    }[];
};

export function serialiseScheme(scheme: Scheme): SchemeSerialised {
    const variables = Array.from(scheme.variables);
    const order = Array.from(scheme.order.entries()).map(([variableFrom, variableTos]) => ({
        variableFrom,
        variableTos: Array.from(variableTos),
    }));

    const statements = Array.from(scheme.statements.entries()).flatMap(([variableFrom, variableToStatements]) => 
        Array.from(variableToStatements.entries()).map(([variableTo, statements]) => ({
            variableFrom,
            variableTo,
            statements
        }))
    );

    return {
        variables,
        order,
        statements,
    };
}

export function deserialiseScheme(data: SchemeSerialised): Scheme {
    const variables = new Set(data.variables);
    const order = new Map(data.order.map(({ variableFrom, variableTos }) => [variableFrom, new Set(variableTos)]));
    const statements = new Map<string, Map<string, Statement[]>>();

    for (const { variableFrom, variableTo, statements: stmt } of data.statements) {
        if (!statements.has(variableFrom)) {
            statements.set(variableFrom, new Map());
        }
        statements.get(variableFrom)!.set(variableTo, stmt);
    }

    return {
        variables,
        order,
        statements,
    };
}
